crates/local-cluster-runner/src/node/mod.rs

```diff
@@ -0,0 +1,740 @@
+                    })
+                    .ok_or(BinarySourceError::NotACargoTest)?;
+
+                let bin = test_bin_dir.join("restate-server");
+                if !bin.exists() {
+                    return Err(BinarySourceError::NotACargoTest);
+                }
+
+                Ok(bin.into())
+            }
+        }
+    }
+}
+
+pub struct StartedNode {
+    status: StartedNodeStatus,
+    config: Configuration,
+}
+
+enum StartedNodeStatus {
+    Running {
+        child_handle: JoinHandle<Result<ExitStatus, io::Error>>,
+        searcher: Searcher,
+        pid: u32,
+    },
+    Exited(ExitStatus),
+    Failed(ErrorKind),
+}
+
+impl Future for StartedNodeStatus {
+    type Output = Result<ExitStatus, io::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match &mut *self {
+            StartedNodeStatus::Running { child_handle, .. } => match child_handle.poll_unpin(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(Ok(Ok(status))) => {
+                    *self = StartedNodeStatus::Exited(status);
+                    Poll::Ready(Ok(status))
+                }
+                Poll::Ready(Ok(Err(err))) => {
+                    *self = StartedNodeStatus::Failed(err.kind());
+                    Poll::Ready(Err(err))
+                }
+                Poll::Ready(Err(err)) => panic!("Join error on started node task: {err}"),
+            },
+            StartedNodeStatus::Exited(status) => Poll::Ready(Ok(*status)),
+            StartedNodeStatus::Failed(error_kind) => Poll::Ready(Err((*error_kind).into())),
+        }
+    }
+}
+
+impl StartedNode {
+    pub async fn kill(&mut self) -> io::Result<ExitStatus> {
+        match &mut self.status {
+            StartedNodeStatus::Exited(status) => Ok(*status),
+            StartedNodeStatus::Failed(kind) => Err((*kind).into()),
+            StartedNodeStatus::Running { pid, .. } => {
+                info!(
+                    "Sending SIGKILL to node {} (pid {})",
+                    self.config.node_name(),
+                    pid
+                );
+                let code = unsafe { libc::kill(*pid as libc::pid_t, libc::SIGKILL) };
+                if code == -1 {
+                    Err(io::Error::last_os_error())
+                } else {
+                    (&mut self.status).await
+                }
+            }
+        }
+    }
+
+    pub fn terminate(&self) -> io::Result<()> {
+        match self.status {
+            StartedNodeStatus::Exited(_) => Ok(()),
+            StartedNodeStatus::Failed(kind) => Err((kind).into()),
+            StartedNodeStatus::Running { pid, .. } => {
+                info!(
+                    "Sending SIGTERM to node {} (pid {})",
+                    self.config.node_name(),
+                    pid
+                );
+                let code = unsafe { libc::kill(pid as libc::pid_t, libc::SIGTERM) };
```

Comment:
Any particular reason against using `nix` crate for those?

Reply:
the main reason is i didnt know about it :D